### ISSUE TITLE
[rccl] Build (linking) fails with llvm crashing

### ISSUE THREAD

**Issue by chengyuhui** [rocm-arch/rocm-arch#302]
All other relevant dependencies are installed from arch4edu. It seems that at least one other user on AUR has the same problem.

Sorry for these Chinese messages, but they probably shouldn't matter.

<details>
  <summary>Log</summary>

  ```
-- The CXX compiler identification is Clang 11.0.0
-- Check for working CXX compiler: /opt/rocm/hip/bin/hipcc
-- Check for working CXX compiler: /opt/rocm/hip/bin/hipcc - works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Performing Test HIP_CLANG_SUPPORTS_PARALLEL_JOBS
-- Performing Test HIP_CLANG_SUPPORTS_PARALLEL_JOBS - Success
-- HIP compiler: clang
-- HIP runtime: ROCclr
-- Configuring done
-- Generating done
-- Build files have been written to: /tmp/makepkg/rccl/src/build
make: 进入目录“/tmp/makepkg/rccl/src/build”
make[1]: 进入目录“/tmp/makepkg/rccl/src/build”
make[2]: 进入目录“/tmp/makepkg/rccl/src/build”
Scanning dependencies of target rccl
make[2]: 离开目录“/tmp/makepkg/rccl/src/build”
make[2]: 进入目录“/tmp/makepkg/rccl/src/build”
[  2%] Building CXX object CMakeFiles/rccl.dir/src/collectives/device/reduce.cpp.o
[  5%] Building CXX object CMakeFiles/rccl.dir/src/collectives/device/all_reduce.cpp.o
[  8%] Building CXX object CMakeFiles/rccl.dir/src/collectives/device/all_gather.cpp.o
[ 10%] Building CXX object CMakeFiles/rccl.dir/src/collectives/device/reduce_scatter.cpp.o
[ 13%] Building CXX object CMakeFiles/rccl.dir/src/collectives/device/broadcast.cpp.o
[ 16%] Building CXX object CMakeFiles/rccl.dir/src/graph/paths.cc.o
[ 18%] Building CXX object CMakeFiles/rccl.dir/src/graph/search.cc.o
[ 21%] Building CXX object CMakeFiles/rccl.dir/src/graph/xml.cc.o
[ 24%] Building CXX object CMakeFiles/rccl.dir/src/collectives/device/functions.cpp.o
[ 27%] Building CXX object CMakeFiles/rccl.dir/src/graph/topo.cc.o
[ 29%] Building CXX object CMakeFiles/rccl.dir/src/init.cc.o
[ 32%] Building CXX object CMakeFiles/rccl.dir/src/graph/trees.cc.o
[ 35%] Building CXX object CMakeFiles/rccl.dir/src/graph/rings.cc.o
[ 37%] Building CXX object CMakeFiles/rccl.dir/src/collectives/all_gather_api.cc.o
[ 40%] Building CXX object CMakeFiles/rccl.dir/src/graph/connect.cc.o
[ 43%] Building CXX object CMakeFiles/rccl.dir/src/graph/tuning.cc.o
[ 45%] Building CXX object CMakeFiles/rccl.dir/src/collectives/all_reduce_api.cc.o
[ 48%] Building CXX object CMakeFiles/rccl.dir/src/collectives/reduce_api.cc.o
[ 51%] Building CXX object CMakeFiles/rccl.dir/src/collectives/broadcast_api.cc.o
[ 54%] Building CXX object CMakeFiles/rccl.dir/src/collectives/reduce_scatter_api.cc.o
[ 56%] Building CXX object CMakeFiles/rccl.dir/src/channel.cc.o
[ 59%] Building CXX object CMakeFiles/rccl.dir/src/misc/argcheck.cc.o
[ 62%] Building CXX object CMakeFiles/rccl.dir/src/misc/nvmlwrap_stub.cc.o
[ 64%] Building CXX object CMakeFiles/rccl.dir/src/misc/utils.cc.o
[ 67%] Building CXX object CMakeFiles/rccl.dir/src/misc/ibvwrap.cc.o
[ 70%] Building CXX object CMakeFiles/rccl.dir/src/transport/coll_net.cc.o
[ 72%] Building CXX object CMakeFiles/rccl.dir/src/transport/net.cc.o
[ 75%] Building CXX object CMakeFiles/rccl.dir/src/transport/net_ib.cc.o
[ 78%] Building CXX object CMakeFiles/rccl.dir/src/transport/net_socket.cc.o
[ 81%] Building CXX object CMakeFiles/rccl.dir/src/transport/p2p.cc.o
[ 83%] Building CXX object CMakeFiles/rccl.dir/src/transport/shm.cc.o
[ 86%] Building CXX object CMakeFiles/rccl.dir/src/transport.cc.o
[ 89%] Building CXX object CMakeFiles/rccl.dir/src/debug.cc.o
[ 91%] Building CXX object CMakeFiles/rccl.dir/src/group.cc.o
[ 94%] Building CXX object CMakeFiles/rccl.dir/src/bootstrap.cc.o
[ 97%] Building CXX object CMakeFiles/rccl.dir/src/enqueue.cc.o
[100%] Linking CXX shared library librccl.so
clang-11: warning: argument unused during compilation: '--hip-device-lib-path=/opt/rocm/lib' [-Wunused-command-line-argument]
Call parameter type does not match function signature!
  %StackGuardSlot = alloca i8*, addrspace(5)
 i8**  call void @llvm.stackprotector(i8* Call parameter type does not match function signature!
  %StackGuardSlot = alloca i8*, addrspace(5)
 i8**  call void @llvm.stackprotector(i8* %2, i8* addrspace(5)* %StackGuardSlot)
in function _Z25ncclAllReduceTreeLLKernelILi2E7FuncSumIaEaEvP14CollectiveArgs
LLVM ERROR: Broken function found, compilation aborted!
PLEASE submit a bug report to https://bugs.llvm.org/ and include the crash backtrace.
Stack dump:
0.      Program arguments: /opt/rocm/llvm/bin/llc /tmp/all_reduce-09c340-gfx906-optimized-a09294.bc -O2 -mtriple=amdgcn-amd-amdhsa -mcpu=gfx906 -filetype=obj -o /tmp/all_reduce-09c340-gfx906-064672.o 
1.      Running pass 'CallGraph Pass Manager' on module '/tmp/all_reduce-09c340-gfx906-optimized-a09294.bc'.
2.      Running pass 'Module Verifier' on function '@_Z25ncclAllReduceTreeLLKernelILi2E7FuncSumIaEaEvP14CollectiveArgs'
 #0 0x000056269a57722b llvm::sys::PrintStackTrace(llvm::raw_ostream&) (/opt/rocm/llvm/bin/llc+0x191b22b)
 #1 0x000056269a574de4 llvm::sys::RunSignalHandlers() (/opt/rocm/llvm/bin/llc+0x1918de4)
 #2 0x000056269a574f39 (/opt/rocm/llvm/bin/llc+0x1918f39)
 #3 0x00007fb38a1ce960 __restore_rt (/usr/lib/libpthread.so.0+0x14960)
 #4 0x00007fb389c60355 raise (/usr/lib/libc.so.6+0x3c355)
 #5 0x00007fb389c49853 abort (/usr/lib/libc.so.6+0x25853)
 #6 0x000056269a4f47e4 llvm::report_fatal_error(llvm::Twine const&, bool) (/opt/rocm/llvm/bin/llc+0x18987e4)
 #7 0x000056269a4f4909 (/opt/rocm/llvm/bin/llc+0x1898909)
 #8 0x0000562699ec78a8 (/opt/rocm/llvm/bin/llc+0x126b8a8)
 #9 0x0000562699e3fad0 llvm::FPPassManager::runOnFunction(llvm::Function&) (/opt/rocm/llvm/bin/llc+0x11e3ad0)
#10 0x00005626996d43a1 (/opt/rocm/llvm/bin/llc+0xa783a1)
#11 0x0000562699e41480 llvm::legacy::PassManagerImpl::run(llvm::Module&) (/opt/rocm/llvm/bin/llc+0x11e5480)
#12 0x0000562699096ead main (/opt/rocm/llvm/bin/llc+0x43aead)
#13 0x00007fb389c4b002 __libc_start_main (/usr/lib/libc.so.6+0x27002)
#14 0x00005626990f489e _start (/opt/rocm/llvm/bin/llc+0x49889e)
%2, i8* addrspace(5)* %StackGuardSlot)
in function _Z25ncclAllReduceTreeLLKernelILi2E7FuncSumIaEaEvP14CollectiveArgs
LLVM ERROR: Broken function found, compilation aborted!
PLEASE submit a bug report to https://bugs.llvm.org/ and include the crash backtrace.
Stack dump:
0.      Program arguments: /opt/rocm/llvm/bin/llc /tmp/all_reduce-6db8b3-gfx803-optimized-8e3340.bc -O2 -mtriple=amdgcn-amd-amdhsa -mcpu=gfx803 -filetype=obj -o /tmp/all_reduce-6db8b3-gfx803-ccf512.o 
1.      Running pass 'CallGraph Pass Manager' on module '/tmp/all_reduce-6db8b3-gfx803-optimized-8e3340.bc'.
2.      Running pass 'Module Verifier' on function '@_Z25ncclAllReduceTreeLLKernelILi2E7FuncSumIaEaEvP14CollectiveArgs'
 #0 0x000055a89e6c522b llvm::sys::PrintStackTrace(llvm::raw_ostream&) (/opt/rocm/llvm/bin/llc+0x191b22b)
 #1 0x000055a89e6c2de4 llvm::sys::RunSignalHandlers() (/opt/rocm/llvm/bin/llc+0x1918de4)
 #2 0x000055a89e6c2f39 (/opt/rocm/llvm/bin/llc+0x1918f39)
 #3 0x00007f58422ae960 __restore_rt (/usr/lib/libpthread.so.0+0x14960)
 #4 0x00007f5841d40355 raise (/usr/lib/libc.so.6+0x3c355)
 #5 0x00007f5841d29853 abort (/usr/lib/libc.so.6+0x25853)
 #6 0x000055a89e6427e4 llvm::report_fatal_error(llvm::Twine const&, bool) (/opt/rocm/llvm/bin/llc+0x18987e4)
 #7 0x000055a89e642909 (/opt/rocm/llvm/bin/llc+0x1898909)
 #8 0x000055a89e0158a8 (/opt/rocm/llvm/bin/llc+0x126b8a8)
 #9 0x000055a89df8dad0 llvm::FPPassManager::runOnFunction(llvm::Function&) (/opt/rocm/llvm/bin/llc+0x11e3ad0)
#10 0x000055a89d8223a1 (/opt/rocm/llvm/bin/llc+0xa783a1)
#11 0x000055a89df8f480 llvm::legacy::PassManagerImpl::run(llvm::Module&) (/opt/rocm/llvm/bin/llc+0x11e5480)
#12 0x000055a89d1e4ead main (/opt/rocm/llvm/bin/llc+0x43aead)
#13 0x00007f5841d2b002 __libc_start_main (/usr/lib/libc.so.6+0x27002)
#14 0x000055a89d24289e _start (/opt/rocm/llvm/bin/llc+0x49889e)
Call parameter type does not match function signature!
  %StackGuardSlot = alloca i8*, addrspace(5)
 i8**  call void @llvm.stackprotector(i8* %2, i8* addrspace(5)* %StackGuardSlot)
in function _Z25ncclAllReduceTreeLLKernelILi2E7FuncSumIaEaEvP14CollectiveArgs
LLVM ERROR: Broken function found, compilation aborted!
PLEASE submit a bug report to https://bugs.llvm.org/ and include the crash backtrace.
Stack dump:
0.      Program arguments: /opt/rocm/llvm/bin/llc /tmp/all_reduce-2ea6c7-gfx900-optimized-d060d9.bc -O2 -mtriple=amdgcn-amd-amdhsa -mcpu=gfx900 -filetype=obj -o /tmp/all_reduce-2ea6c7-gfx900-05838c.o 
1.      Running pass 'CallGraph Pass Manager' on module '/tmp/all_reduce-2ea6c7-gfx900-optimized-d060d9.bc'.
2.      Running pass 'Module Verifier' on function '@_Z25ncclAllReduceTreeLLKernelILi2E7FuncSumIaEaEvP14CollectiveArgs'
 #0 0x00005653b829922b llvm::sys::PrintStackTrace(llvm::raw_ostream&) (/opt/rocm/llvm/bin/llc+0x191b22b)
 #1 0x00005653b8296de4 llvm::sys::RunSignalHandlers() (/opt/rocm/llvm/bin/llc+0x1918de4)
 #2 0x00005653b8296f39 (/opt/rocm/llvm/bin/llc+0x1918f39)
 #3 0x00007efe523b8960 __restore_rt (/usr/lib/libpthread.so.0+0x14960)
 #4 0x00007efe51e4a355 raise (/usr/lib/libc.so.6+0x3c355)
 #5 0x00007efe51e33853 abort (/usr/lib/libc.so.6+0x25853)
 #6 0x00005653b82167e4 llvm::report_fatal_error(llvm::Twine const&, bool) (/opt/rocm/llvm/bin/llc+0x18987e4)
 #7 0x00005653b8216909 (/opt/rocm/llvm/bin/llc+0x1898909)
 #8 0x00005653b7be98a8 (/opt/rocm/llvm/bin/llc+0x126b8a8)
 #9 0x00005653b7b61ad0 llvm::FPPassManager::runOnFunction(llvm::Function&) (/opt/rocm/llvm/bin/llc+0x11e3ad0)
#10 0x00005653b73f63a1 (/opt/rocm/llvm/bin/llc+0xa783a1)
#11 0x00005653b7b63480 llvm::legacy::PassManagerImpl::run(llvm::Module&) (/opt/rocm/llvm/bin/llc+0x11e5480)
#12 0x00005653b6db8ead main (/opt/rocm/llvm/bin/llc+0x43aead)
#13 0x00007efe51e35002 __libc_start_main (/usr/lib/libc.so.6+0x27002)
#14 0x00005653b6e1689e _start (/opt/rocm/llvm/bin/llc+0x49889e)
clang-11: error: unable to execute command: Aborted (core dumped)
clang-11: error: unable to execute command: Aborted (core dumped)
Call parameter type does not match function signature!
  %StackGuardSlot = alloca i8*, addrspace(5)
 i8**  call void @llvm.stackprotector(i8* %2, i8* addrspace(5)* %StackGuardSlot)
in function _Z25ncclAllReduceTreeLLKernelILi2E7FuncSumIaEaEvP14CollectiveArgs
LLVM ERROR: Broken function found, compilation aborted!
PLEASE submit a bug report to https://bugs.llvm.org/ and include the crash backtrace.
Stack dump:
0.      Program arguments: /opt/rocm/llvm/bin/llc /tmp/all_reduce-5f47e3-gfx908-optimized-779e04.bc -O2 -mtriple=amdgcn-amd-amdhsa -mcpu=gfx908 -filetype=obj -o /tmp/all_reduce-5f47e3-gfx908-ab53de.o 
1.      Running pass 'CallGraph Pass Manager' on module '/tmp/all_reduce-5f47e3-gfx908-optimized-779e04.bc'.
2.      Running pass 'Module Verifier' on function '@_Z25ncclAllReduceTreeLLKernelILi2E7FuncSumIaEaEvP14CollectiveArgs'
 #0 0x000055672fe9922b llvm::sys::PrintStackTrace(llvm::raw_ostream&) (/opt/rocm/llvm/bin/llc+0x191b22b)
 #1 0x000055672fe96de4 llvm::sys::RunSignalHandlers() (/opt/rocm/llvm/bin/llc+0x1918de4)
 #2 0x000055672fe96f39 (/opt/rocm/llvm/bin/llc+0x1918f39)
 #3 0x00007f378b6ba960 __restore_rt (/usr/lib/libpthread.so.0+0x14960)
 #4 0x00007f378b14c355 raise (/usr/lib/libc.so.6+0x3c355)
 #5 0x00007f378b135853 abort (/usr/lib/libc.so.6+0x25853)
 #6 0x000055672fe167e4 llvm::report_fatal_error(llvm::Twine const&, bool) (/opt/rocm/llvm/bin/llc+0x18987e4)
 #7 0x000055672fe16909 (/opt/rocm/llvm/bin/llc+0x1898909)
 #8 0x000055672f7e98a8 (/opt/rocm/llvm/bin/llc+0x126b8a8)
 #9 0x000055672f761ad0 llvm::FPPassManager::runOnFunction(llvm::Function&) (/opt/rocm/llvm/bin/llc+0x11e3ad0)
#10 0x000055672eff63a1 (/opt/rocm/llvm/bin/llc+0xa783a1)
#11 0x000055672f763480 llvm::legacy::PassManagerImpl::run(llvm::Module&) (/opt/rocm/llvm/bin/llc+0x11e5480)
#12 0x000055672e9b8ead main (/opt/rocm/llvm/bin/llc+0x43aead)
#13 0x00007f378b137002 __libc_start_main (/usr/lib/libc.so.6+0x27002)
#14 0x000055672ea1689e _start (/opt/rocm/llvm/bin/llc+0x49889e)
clang-11: error: unable to execute command: Aborted (core dumped)
clang-11: error: unable to execute command: Aborted (core dumped)
clang-11: error: amdgcn-link command failed due to signal (use -v to see invocation)
clang-11: error: amdgcn-link command failed due to signal (use -v to see invocation)
clang-11: error: amdgcn-link command failed due to signal (use -v to see invocation)
clang-11: error: amdgcn-link command failed due to signal (use -v to see invocation)
clang version 11.0.0 
Target: x86_64-pc-linux-gnu
Thread model: posix
InstalledDir: /opt/rocm/llvm/bin
clang-11: note: diagnostic msg: 
********************

PLEASE ATTACH THE FOLLOWING FILES TO THE BUG REPORT:
Preprocessed source(s) and associated run script(s) are located at:
clang-11: note: diagnostic msg: /tmp/libamdhip64-50dadf.c
clang-11: note: diagnostic msg: /tmp/libamdhip64-50dadf.sh
clang-11: note: diagnostic msg: 

********************
make[2]: *** [CMakeFiles/rccl.dir/build.make:630：librccl.so.1.0] 错误 254
make[2]: 离开目录“/tmp/makepkg/rccl/src/build”
make[1]: *** [CMakeFiles/Makefile2:96：CMakeFiles/rccl.dir/all] 错误 2
make[1]: 离开目录“/tmp/makepkg/rccl/src/build”
make: *** [Makefile:172：all] 错误 2
make: 离开目录“/tmp/makepkg/rccl/src/build”
  ```
</details>

[libamdhip64-50dadf.tar.gz](https://github.com/rocm-arch/rocm-arch/files/4941483/libamdhip64-50dadf.tar.gz)

**Comment by tpkessler** [comment]
Hey @chengyuhui! To get English error messages, set `LANG=C`:

```bash
$ LANG=C makepkg
```

I cannot reproduce your build error (all packages installed from `arch4edu`). It's likely a compiler and not a packaging bug.

**Comment by axiaobugs** [comment]
Same error with me.
=> Making package: rccl 3.5.0-2 (Sun 19 Jul 2020 14:41:34)
==> Checking runtime dependencies...
==> Checking buildtime dependencies...
==> Retrieving sources...
  -> Found rccl-3.5.0.tar.gz
==> Validating source files with sha256sums...
    rccl-3.5.0.tar.gz ... Passed
==> Extracting sources...
  -> Extracting rccl-3.5.0.tar.gz with bsdtar
==> Removing existing $pkgdir/ directory...
==> Starting build()...
-- HIP compiler: clang
-- HIP runtime: ROCclr
-- Configuring done
-- Generating done
-- Build files have been written to: /home/axiaobug/pamac-build/rccl/src/build
make: Nothing to be done for 'build'.
==> Entering fakeroot environment...
==> Starting package()...
make: Entering directory '/home/axiaobug/pamac-build/rccl/src/build'
make[1]: Entering directory '/home/axiaobug/pamac-build/rccl/src/build'
make[2]: Entering directory '/home/axiaobug/pamac-build/rccl/src/build'
make[2]: Leaving directory '/home/axiaobug/pamac-build/rccl/src/build'
make[2]: Entering directory '/home/axiaobug/pamac-build/rccl/src/build'
[  2%] Linking CXX shared library librccl.so
HIP_PATH=/opt/rocm/hip
HIP_PLATFORM=hcc
HIP_COMPILER=clang
HIP_RUNTIME=ROCclr
HIP_ROCCLR_HOME=/opt/rocm/hip
HIP_CLANG_PATH=/opt/rocm/llvm/bin
HIP_CLANG_INCLUDE_PATH=/opt/rocm/llvm/lib/clang/11.0.0/include
HIP_INCLUDE_PATH=/opt/rocm/hip/include
HIP_LIB_PATH=/opt/rocm/hip/lib
DEVICE_LIB_PATH=/opt/rocm/lib
hipcc-args: -fPIC -D_FORTIFY_SOURCE=2 -march=x86-64 -mtune=generic -O2 -pipe -fstack-protector-strong -fno-plt -Wl,-O1,--sort-common,--as-needed,-z,relro,-z,now -shared -Wl,-soname,librccl.so.1 -o librccl.so.1.0 CMakeFiles/rccl.dir/src/collectives/device/all_reduce.cpp.o CMakeFiles/rccl.dir/src/collectives/device/all_gather.cpp.o CMakeFiles/rccl.dir/src/collectives/device/reduce.cpp.o CMakeFiles/rccl.dir/src/collectives/device/broadcast.cpp.o CMakeFiles/rccl.dir/src/collectives/device/reduce_scatter.cpp.o CMakeFiles/rccl.dir/src/collectives/device/functions.cpp.o CMakeFiles/rccl.dir/src/init.cc.o CMakeFiles/rccl.dir/src/graph/trees.cc.o CMakeFiles/rccl.dir/src/graph/rings.cc.o CMakeFiles/rccl.dir/src/graph/paths.cc.o CMakeFiles/rccl.dir/src/graph/search.cc.o CMakeFiles/rccl.dir/src/graph/connect.cc.o CMakeFiles/rccl.dir/src/graph/tuning.cc.o CMakeFiles/rccl.dir/src/graph/topo.cc.o CMakeFiles/rccl.dir/src/graph/xml.cc.o CMakeFiles/rccl.dir/src/collectives/all_reduce_api.cc.o CMakeFiles/rccl.dir/src/collectives/all_gather_api.cc.o CMakeFiles/rccl.dir/src/collectives/reduce_api.cc.o CMakeFiles/rccl.dir/src/collectives/broadcast_api.cc.o CMakeFiles/rccl.dir/src/collectives/reduce_scatter_api.cc.o CMakeFiles/rccl.dir/src/channel.cc.o CMakeFiles/rccl.dir/src/misc/argcheck.cc.o CMakeFiles/rccl.dir/src/misc/nvmlwrap_stub.cc.o CMakeFiles/rccl.dir/src/misc/utils.cc.o CMakeFiles/rccl.dir/src/misc/ibvwrap.cc.o CMakeFiles/rccl.dir/src/transport/coll_net.cc.o CMakeFiles/rccl.dir/src/transport/net.cc.o CMakeFiles/rccl.dir/src/transport/net_ib.cc.o CMakeFiles/rccl.dir/src/transport/net_socket.cc.o CMakeFiles/rccl.dir/src/transport/p2p.cc.o CMakeFiles/rccl.dir/src/transport/shm.cc.o CMakeFiles/rccl.dir/src/transport.cc.o CMakeFiles/rccl.dir/src/debug.cc.o CMakeFiles/rccl.dir/src/group.cc.o CMakeFiles/rccl.dir/src/bootstrap.cc.o CMakeFiles/rccl.dir/src/enqueue.cc.o --amdgpu-target=gfx803 --amdgpu-target=gfx900 --amdgpu-target=gfx906 --amdgpu-target=gfx908 -fgpu-rdc -parallel-jobs=4 /opt/rocm/hip/lib/libamdhip64.so.3.5.20261- --hip-device-lib-path=/opt/rocm/lib --hip-link --cuda-gpu-arch=gfx803 --cuda-gpu-arch=gfx900 --cuda-gpu-arch=gfx906 --cuda-gpu-arch=gfx908
hipcc-cmd: /opt/rocm/llvm/bin/clang++  -L/opt/rocm/hip/lib --hip-link --cuda-gpu-arch=gfx803 --cuda-gpu-arch=gfx900 --cuda-gpu-arch=gfx906 --cuda-gpu-arch=gfx908 -lgcc_s -lgcc -lpthread -lm  -fPIC -D\_FORTIFY\_SOURCE=2 -march=x86-64 -mtune=generic -O2 -pipe -fstack-protector-strong -fno-plt -Wl,-O1,--sort-common,--as-needed,-z,relro,-z,now -shared -Wl,-soname,librccl.so.1 -o librccl.so.1.0 CMakeFiles\/rccl.dir\/src\/collectives\/device\/all\_reduce.cpp.o CMakeFiles\/rccl.dir\/src\/collectives\/device\/all\_gather.cpp.o CMakeFiles\/rccl.dir\/src\/collectives\/device\/reduce.cpp.o CMakeFiles\/rccl.dir\/src\/collectives\/device\/broadcast.cpp.o CMakeFiles\/rccl.dir\/src\/collectives\/device\/reduce\_scatter.cpp.o CMakeFiles\/rccl.dir\/src\/collectives\/device\/functions.cpp.o CMakeFiles\/rccl.dir\/src\/init.cc.o CMakeFiles\/rccl.dir\/src\/graph\/trees.cc.o CMakeFiles\/rccl.dir\/src\/graph\/rings.cc.o CMakeFiles\/rccl.dir\/src\/graph\/paths.cc.o CMakeFiles\/rccl.dir\/src\/graph\/search.cc.o CMakeFiles\/rccl.dir\/src\/graph\/connect.cc.o CMakeFiles\/rccl.dir\/src\/graph\/tuning.cc.o CMakeFiles\/rccl.dir\/src\/graph\/topo.cc.o CMakeFiles\/rccl.dir\/src\/graph\/xml.cc.o CMakeFiles\/rccl.dir\/src\/collectives\/all\_reduce\_api.cc.o CMakeFiles\/rccl.dir\/src\/collectives\/all\_gather\_api.cc.o CMakeFiles\/rccl.dir\/src\/collectives\/reduce\_api.cc.o CMakeFiles\/rccl.dir\/src\/collectives\/broadcast\_api.cc.o CMakeFiles\/rccl.dir\/src\/collectives\/reduce\_scatter\_api.cc.o CMakeFiles\/rccl.dir\/src\/channel.cc.o CMakeFiles\/rccl.dir\/src\/misc\/argcheck.cc.o CMakeFiles\/rccl.dir\/src\/misc\/nvmlwrap\_stub.cc.o CMakeFiles\/rccl.dir\/src\/misc\/utils.cc.o CMakeFiles\/rccl.dir\/src\/misc\/ibvwrap.cc.o CMakeFiles\/rccl.dir\/src\/transport\/coll\_net.cc.o CMakeFiles\/rccl.dir\/src\/transport\/net.cc.o CMakeFiles\/rccl.dir\/src\/transport\/net\_ib.cc.o CMakeFiles\/rccl.dir\/src\/transport\/net\_socket.cc.o CMakeFiles\/rccl.dir\/src\/transport\/p2p.cc.o CMakeFiles\/rccl.dir\/src\/transport\/shm.cc.o CMakeFiles\/rccl.dir\/src\/transport.cc.o CMakeFiles\/rccl.dir\/src\/debug.cc.o CMakeFiles\/rccl.dir\/src\/group.cc.o CMakeFiles\/rccl.dir\/src\/bootstrap.cc.o CMakeFiles\/rccl.dir\/src\/enqueue.cc.o -fgpu-rdc -parallel-jobs=4 \/opt\/rocm\/hip\/lib\/libamdhip64.so.3.5.20261- --hip-device-lib-path=\/opt\/rocm\/lib --hip-link --cuda-gpu-arch=gfx803 --cuda-gpu-arch=gfx900 --cuda-gpu-arch=gfx906 --cuda-gpu-arch=gfx908 -Wl,--enable-new-dtags -Wl,--rpath=/opt/rocm/hip/lib:/opt/rocm/lib -lhip_hcc 
clang-11: warning: argument unused during compilation: '--hip-device-lib-path=/opt/rocm/lib' [-Wunused-command-line-argument]
Call parameter type does not match function signature!
  %StackGuardSlot = alloca i8*, addrspace(5)
 i8**  call void @llvm.stackprotector(i8* %2, i8* addrspace(5)* %StackGuardSlot)
in function _Z25ncclAllReduceTreeLLKernelILi2E7FuncSumIaEaEvP14CollectiveArgs
LLVM ERROR: Broken function found, compilation aborted!
PLEASE submit a bug report to https://bugs.llvm.org/ and include the crash backtrace.
Stack dump:
0.	Program arguments: /opt/rocm/llvm/bin/llc /tmp/all_reduce-ab5dcf-gfx803-optimized-a359a8.bc -O2 -mtriple=amdgcn-amd-amdhsa -mcpu=gfx803 -filetype=obj -o /tmp/all_reduce-ab5dcf-gfx803-3dbdc4.o 
1.	Running pass 'CallGraph Pass Manager' on module '/tmp/all_reduce-ab5dcf-gfx803-optimized-a359a8.bc'.
2.	Running pass 'Module Verifier' on function '@_Z25ncclAllReduceTreeLLKernelILi2E7FuncSumIaEaEvP14CollectiveArgs'
 #0 0x000055c3e634622b llvm::sys::PrintStackTrace(llvm::raw_ostream&) (/opt/rocm/llvm/bin/llc+0x191b22b)
 #1 0x000055c3e6343de4 llvm::sys::RunSignalHandlers() (/opt/rocm/llvm/bin/llc+0x1918de4)
 #2 0x000055c3e6343f39 (/opt/rocm/llvm/bin/llc+0x1918f39)
 #3 0x00007f47ff897960 __restore_rt (/usr/lib/libpthread.so.0+0x14960)
 #4 0x00007f47ff329355 raise (/usr/lib/libc.so.6+0x3c355)
 #5 0x00007f47ff312853 abort (/usr/lib/libc.so.6+0x25853)
 #6 0x000055c3e62c37e4 llvm::report_fatal_error(llvm::Twine const&, bool) (/opt/rocm/llvm/bin/llc+0x18987e4)
 #7 0x000055c3e62c3909 (/opt/rocm/llvm/bin/llc+0x1898909)
 #8 0x000055c3e5c968a8 (/opt/rocm/llvm/bin/llc+0x126b8a8)
 #9 0x000055c3e5c0ead0 llvm::FPPassManager::runOnFunction(llvm::Function&) (/opt/rocm/llvm/bin/llc+0x11e3ad0)
#10 0x000055c3e54a33a1 (/opt/rocm/llvm/bin/llc+0xa783a1)
#11 0x000055c3e5c10480 llvm::legacy::PassManagerImpl::run(llvm::Module&) (/opt/rocm/llvm/bin/llc+0x11e5480)
#12 0x000055c3e4e65ead main (/opt/rocm/llvm/bin/llc+0x43aead)
#13 0x00007f47ff314002 __libc_start_main (/usr/lib/libc.so.6+0x27002)
#14 0x000055c3e4ec389e _start (/opt/rocm/llvm/bin/llc+0x49889e)
clang-11: error: unable to execute command: Aborted (core dumped)
Call parameter type does not match function signature!
  %StackGuardSlot = alloca i8*, addrspace(5)
 i8**  call void @llvm.stackprotector(i8* %2, i8* addrspace(5)* %StackGuardSlot)
in function _Z25ncclAllReduceTreeLLKernelILi2E7FuncSumIaEaEvP14CollectiveArgs
LLVM ERROR: Broken function found, compilation aborted!
PLEASE submit a bug report to https://bugs.llvm.org/ and include the crash backtrace.
Stack dump:
0.	Program arguments: /opt/rocm/llvm/bin/llc /tmp/all_reduce-57a18b-gfx908-optimized-f6adc2.bc -O2 -mtriple=amdgcn-amd-amdhsa -mcpu=gfx908 -filetype=obj -o /tmp/all_reduce-57a18b-gfx908-cc4ebd.o 
1.	Running pass 'CallGraph Pass Manager' on module '/tmp/all_reduce-57a18b-gfx908-optimized-f6adc2.bc'.
2.	Running pass 'Module Verifier' on function '@_Z25ncclAllReduceTreeLLKernelILi2E7FuncSumIaEaEvP14CollectiveArgs'
 #0 0x000055579019f22b llvm::sys::PrintStackTrace(llvm::raw_ostream&) (/opt/rocm/llvm/bin/llc+0x191b22b)
 #1 0x000055579019cde4 llvm::sys::RunSignalHandlers() (/opt/rocm/llvm/bin/llc+0x1918de4)
 #2 0x000055579019cf39 (/opt/rocm/llvm/bin/llc+0x1918f39)
 #3 0x00007efca6c25960 __restore_rt (/usr/lib/libpthread.so.0+0x14960)
 #4 0x00007efca66b7355 raise (/usr/lib/libc.so.6+0x3c355)
 #5 0x00007efca66a0853 abort (/usr/lib/libc.so.6+0x25853)
 #6 0x000055579011c7e4 llvm::report_fatal_error(llvm::Twine const&, bool) (/opt/rocm/llvm/bin/llc+0x18987e4)
 #7 0x000055579011c909 (/opt/rocm/llvm/bin/llc+0x1898909)
 #8 0x000055578faef8a8 (/opt/rocm/llvm/bin/llc+0x126b8a8)
 #9 0x000055578fa67ad0 llvm::FPPassManager::runOnFunction(llvm::Function&) (/opt/rocm/llvm/bin/llc+0x11e3ad0)
#10 0x000055578f2fc3a1 (/opt/rocm/llvm/bin/llc+0xa783a1)
#11 0x000055578fa69480 llvm::legacy::PassManagerImpl::run(llvm::Module&) (/opt/rocm/llvm/bin/llc+0x11e5480)
#12 0x000055578ecbeead main (/opt/rocm/llvm/bin/llc+0x43aead)
#13 0x00007efca66a2002 __libc_start_main (/usr/lib/libc.so.6+0x27002)
#14 0x000055578ed1c89e _start (/opt/rocm/llvm/bin/llc+0x49889e)
Call parameter type does not match function signature!
  %StackGuardSlot = alloca i8*, addrspace(5)
 i8**  call void @llvm.stackprotector(i8* Call parameter type does not match function signature!
  %StackGuardSlot = alloca i8*, addrspace(5)
 i8**  call void @llvm.stackprotector(i8* %2, i8* addrspace(5)* %StackGuardSlot)
in function _Z25ncclAllReduceTreeLLKernelILi2E7FuncSumIaEaEvP14CollectiveArgs
LLVM ERROR: Broken function found, compilation aborted!
PLEASE submit a bug report to https://bugs.llvm.org/ and include the crash backtrace.
Stack dump:
0.	Program arguments: /opt/rocm/llvm/bin/llc /tmp/all_reduce-bf39f1-gfx900-optimized-190b72.bc -O2 -mtriple=amdgcn-amd-amdhsa -mcpu=gfx900 -filetype=obj -o /tmp/all_reduce-bf39f1-gfx900-b150af.o 
1.	Running pass 'CallGraph Pass Manager' on module '/tmp/all_reduce-bf39f1-gfx900-optimized-190b72.bc'.
2.	Running pass 'Module Verifier' on function '@_Z25ncclAllReduceTreeLLKernelILi2E7FuncSumIaEaEvP14CollectiveArgs'
 #0 0x000056348eb3222b llvm::sys::PrintStackTrace(llvm::raw_ostream&) (/opt/rocm/llvm/bin/llc+0x191b22b)
 #1 0x000056348eb2fde4 llvm::sys::RunSignalHandlers() (/opt/rocm/llvm/bin/llc+0x1918de4)
 #2 0x000056348eb2ff39 (/opt/rocm/llvm/bin/llc+0x1918f39)
 #3 0x00007f0f6d633960 __restore_rt (/usr/lib/libpthread.so.0+0x14960)
 #4 0x00007f0f6d0c5355 raise (/usr/lib/libc.so.6+0x3c355)
 #5 0x00007f0f6d0ae853 abort (/usr/lib/libc.so.6+0x25853)
 #6 0x000056348eaaf7e4 llvm::report_fatal_error(llvm::Twine const&, bool) (/opt/rocm/llvm/bin/llc+0x18987e4)
 #7 0x000056348eaaf909 (/opt/rocm/llvm/bin/llc+0x1898909)
 #8 0x000056348e4828a8 (/opt/rocm/llvm/bin/llc+0x126b8a8)
 #9 0x000056348e3faad0 llvm::FPPassManager::runOnFunction(llvm::Function&) (/opt/rocm/llvm/bin/llc+0x11e3ad0)
#10 0x000056348dc8f3a1 (/opt/rocm/llvm/bin/llc+0xa783a1)
#11 0x000056348e3fc480 llvm::legacy::PassManagerImpl::run(llvm::Module&) (/opt/rocm/llvm/bin/llc+0x11e5480)
#12 0x000056348d651ead main (/opt/rocm/llvm/bin/llc+0x43aead)
#13 0x00007f0f6d0b0002 __libc_start_main (/usr/lib/libc.so.6+0x27002)
#14 0x000056348d6af89e _start (/opt/rocm/llvm/bin/llc+0x49889e)
%2, i8* addrspace(5)* %StackGuardSlot)
in function _Z25ncclAllReduceTreeLLKernelILi2E7FuncSumIaEaEvP14CollectiveArgs
LLVM ERROR: Broken function found, compilation aborted!
PLEASE submit a bug report to https://bugs.llvm.org/ and include the crash backtrace.
Stack dump:
0.	Program arguments: /opt/rocm/llvm/bin/llc /tmp/all_reduce-d60fc1-gfx906-optimized-022e95.bc -O2 -mtriple=amdgcn-amd-amdhsa -mcpu=gfx906 -filetype=obj -o /tmp/all_reduce-d60fc1-gfx906-2ae266.o 
1.	Running pass 'CallGraph Pass Manager' on module '/tmp/all_reduce-d60fc1-gfx906-optimized-022e95.bc'.
2.	Running pass 'Module Verifier' on function '@_Z25ncclAllReduceTreeLLKernelILi2E7FuncSumIaEaEvP14CollectiveArgs'
 #0 0x0000562eca26622b llvm::sys::PrintStackTrace(llvm::raw_ostream&) (/opt/rocm/llvm/bin/llc+0x191b22b)
 #1 0x0000562eca263de4 llvm::sys::RunSignalHandlers() (/opt/rocm/llvm/bin/llc+0x1918de4)
 #2 0x0000562eca263f39 (/opt/rocm/llvm/bin/llc+0x1918f39)
 #3 0x00007ff9b0f15960 __restore_rt (/usr/lib/libpthread.so.0+0x14960)
 #4 0x00007ff9b09a7355 raise (/usr/lib/libc.so.6+0x3c355)
 #5 0x00007ff9b0990853 abort (/usr/lib/libc.so.6+0x25853)
 #6 0x0000562eca1e37e4 llvm::report_fatal_error(llvm::Twine const&, bool) (/opt/rocm/llvm/bin/llc+0x18987e4)
 #7 0x0000562eca1e3909 (/opt/rocm/llvm/bin/llc+0x1898909)
 #8 0x0000562ec9bb68a8 (/opt/rocm/llvm/bin/llc+0x126b8a8)
 #9 0x0000562ec9b2ead0 llvm::FPPassManager::runOnFunction(llvm::Function&) (/opt/rocm/llvm/bin/llc+0x11e3ad0)
#10 0x0000562ec93c33a1 (/opt/rocm/llvm/bin/llc+0xa783a1)
#11 0x0000562ec9b30480 llvm::legacy::PassManagerImpl::run(llvm::Module&) (/opt/rocm/llvm/bin/llc+0x11e5480)
#12 0x0000562ec8d85ead main (/opt/rocm/llvm/bin/llc+0x43aead)
#13 0x00007ff9b0992002 __libc_start_main (/usr/lib/libc.so.6+0x27002)
#14 0x0000562ec8de389e _start (/opt/rocm/llvm/bin/llc+0x49889e)
clang-11: error: unable to execute command: Aborted (core dumped)
clang-11: error: unable to execute command: Aborted (core dumped)
clang-11: error: unable to execute command: Aborted (core dumped)
clang-11: error: amdgcn-link command failed due to signal (use -v to see invocation)
clang-11: error: amdgcn-link command failed due to signal (use -v to see invocation)
clang-11: error: amdgcn-link command failed due to signal (use -v to see invocation)
clang-11: error: amdgcn-link command failed due to signal (use -v to see invocation)
clang version 11.0.0 
Target: x86_64-pc-linux-gnu
Thread model: posix
InstalledDir: /opt/rocm/llvm/bin
clang-11: note: diagnostic msg: 
********************

PLEASE ATTACH THE FOLLOWING FILES TO THE BUG REPORT:
Preprocessed source(s) and associated run script(s) are located at:
clang-11: note: diagnostic msg: /tmp/libamdhip64-19a1f8.c
clang-11: note: diagnostic msg: /tmp/libamdhip64-19a1f8.sh
clang-11: note: diagnostic msg: 

********************
make[2]: *** [CMakeFiles/rccl.dir/build.make:630: librccl.so.1.0] Error 254
make[2]: Leaving directory '/home/axiaobug/pamac-build/rccl/src/build'
make[1]: *** [CMakeFiles/Makefile2:96: CMakeFiles/rccl.dir/all] Error 2
make[1]: Leaving directory '/home/axiaobug/pamac-build/rccl/src/build'
make: *** [Makefile:172: all] Error 2
make: Leaving directory '/home/axiaobug/pamac-build/rccl/src/build'
==> ERROR: A failure occurred in package().
    Aborting...

**Comment by tpkessler** [comment]
Can you guys try to compile the latest `llvm-amdgpu` from my PR #305 ? Do your problems persist?

**Comment by chengyuhui** [comment]
I really would like to, but I don't currently have enough resource for compiling such a large package.

**Comment by axiaobugs** [comment]
I am done when I installed nccl

**Comment by acxz** [comment]
@chengyuhui and @axiaobugs are you guys still able to reproduce this error?
The build works just fine for me.

Also I don't understand this statement if you could clarify that would be great.
> I am done when I installed nccl

Does this mean it started working for you after you installed `nccl`?

**Comment by axiaobugs** [comment]
> @chengyuhui and @axiaobugs are you guys still able to reproduce this error?
> The build works just fine for me.
> 
> Also I don't understand this statement if you could clarify that would be great.
> > I am done when I installed nccl
> 
> Does this mean it started working for you after you installed `nccl`?

yes

**Comment by EvilMav** [comment]
Seeing the same thing when compiling  rccl 3.5.0-2 from AUR (using yay). Installation of `nccl` did not change anything.
```
...
[100%] Linking CXX shared library librccl.so
clang-11: warning: argument unused during compilation: '--hip-device-lib-path=/opt/rocm/lib' [-Wunused-command-line-argument]
Call parameter type does not match function signature!
  %StackGuardSlot = alloca i8*, addrspace(5)
 i8**  call void @llvm.stackprotector(i8* %2, i8* addrspace(5)* %StackGuardSlot)
in function _Z25ncclAllReduceTreeLLKernelILi2E7FuncSumIaEaEvP14CollectiveArgs
LLVM ERROR: Broken function found, compilation aborted!
PLEASE submit a bug report to https://bugs.llvm.org/ and include the crash backtrace.
Stack dump:
0.      Program arguments: /opt/rocm/llvm/bin/llc /tmp/all_reduce-761012-gfx803-optimized-e54569.bc -O2 -mtriple=amdgcn-amd-amdhsa -mcpu=gfx803 -filetype=obj -o /tmp/all_reduce-761012-gfx803-3dfb5b.o 
1.      Running pass 'CallGraph Pass Manager' on module '/tmp/all_reduce-761012-gfx803-optimized-e54569.bc'.
2.      Running pass 'Module Verifier' on function '@_Z25ncclAllReduceTreeLLKernelILi2E7FuncSumIaEaEvP14CollectiveArgs'
 #0 0x000055c2fc65122b llvm::sys::PrintStackTrace(llvm::raw_ostream&) (/opt/rocm/llvm/bin/llc+0x191b22b)
 #1 0x000055c2fc64ede4 llvm::sys::RunSignalHandlers() (/opt/rocm/llvm/bin/llc+0x1918de4)
 #2 0x000055c2fc64ef39 (/opt/rocm/llvm/bin/llc+0x1918f39)
 #3 0x00007f940fee0960 __restore_rt (/usr/lib/libpthread.so.0+0x14960)
 #4 0x00007f940f972355 raise (/usr/lib/libc.so.6+0x3c355)
 #5 0x00007f940f95b853 abort (/usr/lib/libc.so.6+0x25853)
 #6 0x000055c2fc5ce7e4 llvm::report_fatal_error(llvm::Twine const&, bool) (/opt/rocm/llvm/bin/llc+0x18987e4)
 #7 0x000055c2fc5ce909 (/opt/rocm/llvm/bin/llc+0x1898909)
 #8 0x000055c2fbfa18a8 (/opt/rocm/llvm/bin/llc+0x126b8a8)
 #9 0x000055c2fbf19ad0 llvm::FPPassManager::runOnFunction(llvm::Function&) (/opt/rocm/llvm/bin/llc+0x11e3ad0)
#10 0x000055c2fb7ae3a1 (/opt/rocm/llvm/bin/llc+0xa783a1)
#11 0x000055c2fbf1b480 llvm::legacy::PassManagerImpl::run(llvm::Module&) (/opt/rocm/llvm/bin/llc+0x11e5480)
#12 0x000055c2fb170ead main (/opt/rocm/llvm/bin/llc+0x43aead)
#13 0x00007f940f95d002 __libc_start_main (/usr/lib/libc.so.6+0x27002)
#14 0x000055c2fb1ce89e _start (/opt/rocm/llvm/bin/llc+0x49889e)
clang-11: error: unable to execute command: Aborted (core dumped)
Call parameter type does not match function signature!
  %StackGuardSlot = alloca i8*, addrspace(5)
 i8**  call void @llvm.stackprotector(i8* %2, i8* addrspace(5)* %StackGuardSlot)
in function _Z25ncclAllReduceTreeLLKernelILi2E7FuncSumIaEaEvP14CollectiveArgs
LLVM ERROR: Broken function found, compilation aborted!
PLEASE submit a bug report to https://bugs.llvm.org/ and include the crash backtrace.
Stack dump:
0.      Program arguments: /opt/rocm/llvm/bin/llc /tmp/all_reduce-0f3c72-gfx900-optimized-93d7ce.bc -O2 -mtriple=amdgcn-amd-amdhsa -mcpu=gfx900 -filetype=obj -o /tmp/all_reduce-0f3c72-gfx900-bb400a.o 
1.      Running pass 'CallGraph Pass Manager' on module '/tmp/all_reduce-0f3c72-gfx900-optimized-93d7ce.bc'.
2.      Running pass 'Module Verifier' on function '@_Z25ncclAllReduceTreeLLKernelILi2E7FuncSumIaEaEvP14CollectiveArgs'
 #0 0x000055bb217e622b llvm::sys::PrintStackTrace(llvm::raw_ostream&) (/opt/rocm/llvm/bin/llc+0x191b22b)
 #1 0x000055bb217e3de4 llvm::sys::RunSignalHandlers() (/opt/rocm/llvm/bin/llc+0x1918de4)
 #2 0x000055bb217e3f39 (/opt/rocm/llvm/bin/llc+0x1918f39)
 #3 0x00007f5e6f840960 __restore_rt (/usr/lib/libpthread.so.0+0x14960)
 #4 0x00007f5e6f2d2355 raise (/usr/lib/libc.so.6+0x3c355)
 #5 0x00007f5e6f2bb853 abort (/usr/lib/libc.so.6+0x25853)
 #6 0x000055bb217637e4 llvm::report_fatal_error(llvm::Twine const&, bool) (/opt/rocm/llvm/bin/llc+0x18987e4)
 #7 0x000055bb21763909 (/opt/rocm/llvm/bin/llc+0x1898909)
 #8 0x000055bb211368a8 (/opt/rocm/llvm/bin/llc+0x126b8a8)
 #9 0x000055bb210aead0 llvm::FPPassManager::runOnFunction(llvm::Function&) (/opt/rocm/llvm/bin/llc+0x11e3ad0)
#10 0x000055bb209433a1 (/opt/rocm/llvm/bin/llc+0xa783a1)
#11 0x000055bb210b0480 llvm::legacy::PassManagerImpl::run(llvm::Module&) (/opt/rocm/llvm/bin/llc+0x11e5480)
#12 0x000055bb20305ead main (/opt/rocm/llvm/bin/llc+0x43aead)
#13 0x00007f5e6f2bd002 __libc_start_main (/usr/lib/libc.so.6+0x27002)
#14 0x000055bb2036389e _start (/opt/rocm/llvm/bin/llc+0x49889e)
Call parameter type does not match function signature!
  %StackGuardSlot = alloca i8*, addrspace(5)
 i8**  call void @llvm.stackprotector(i8* %2, i8* addrspace(5)* %StackGuardSlot)
in function _Z25ncclAllReduceTreeLLKernelILi2E7FuncSumIaEaEvP14CollectiveArgs
LLVM ERROR: Broken function found, compilation aborted!
PLEASE submit a bug report to https://bugs.llvm.org/ and include the crash backtrace.
Stack dump:
0.      Program arguments: /opt/rocm/llvm/bin/llc /tmp/all_reduce-bd0295-gfx908-optimized-6697b4.bc -O2 -mtriple=amdgcn-amd-amdhsa -mcpu=gfx908 -filetype=obj -o /tmp/all_reduce-bd0295-gfx908-7c5088.o 
1.      Running pass 'CallGraph Pass Manager' on module '/tmp/all_reduce-bd0295-gfx908-optimized-6697b4.bc'.
2.      Running pass 'Module Verifier' on function '@_Z25ncclAllReduceTreeLLKernelILi2E7FuncSumIaEaEvP14CollectiveArgs'
 #0 0x0000556536cb022b llvm::sys::PrintStackTrace(llvm::raw_ostream&) (/opt/rocm/llvm/bin/llc+0x191b22b)
 #1 0x0000556536cadde4 llvm::sys::RunSignalHandlers() (/opt/rocm/llvm/bin/llc+0x1918de4)
 #2 0x0000556536cadf39 (/opt/rocm/llvm/bin/llc+0x1918f39)
 #3 0x00007f360bd69960 __restore_rt (/usr/lib/libpthread.so.0+0x14960)
 #4 0x00007f360b7fb355 raise (/usr/lib/libc.so.6+0x3c355)
 #5 0x00007f360b7e4853 abort (/usr/lib/libc.so.6+0x25853)
 #6 0x0000556536c2d7e4 llvm::report_fatal_error(llvm::Twine const&, bool) (/opt/rocm/llvm/bin/llc+0x18987e4)
 #7 0x0000556536c2d909 (/opt/rocm/llvm/bin/llc+0x1898909)
 #8 0x00005565366008a8 (/opt/rocm/llvm/bin/llc+0x126b8a8)
 #9 0x0000556536578ad0 llvm::FPPassManager::runOnFunction(llvm::Function&) (/opt/rocm/llvm/bin/llc+0x11e3ad0)
#10 0x0000556535e0d3a1 (/opt/rocm/llvm/bin/llc+0xa783a1)
#11 0x000055653657a480 llvm::legacy::PassManagerImpl::run(llvm::Module&) (/opt/rocm/llvm/bin/llc+0x11e5480)
#12 0x00005565357cfead main (/opt/rocm/llvm/bin/llc+0x43aead)
#13 0x00007f360b7e6002 __libc_start_main (/usr/lib/libc.so.6+0x27002)
#14 0x000055653582d89e _start (/opt/rocm/llvm/bin/llc+0x49889e)
clang-11: error: unable to execute command: Aborted (core dumped)
clang-11: error: unable to execute command: Aborted (core dumped)
Call parameter type does not match function signature!
  %StackGuardSlot = alloca i8*, addrspace(5)
 i8**  call void @llvm.stackprotector(i8* %2, i8* addrspace(5)* %StackGuardSlot)
in function _Z25ncclAllReduceTreeLLKernelILi2E7FuncSumIaEaEvP14CollectiveArgs
LLVM ERROR: Broken function found, compilation aborted!
PLEASE submit a bug report to https://bugs.llvm.org/ and include the crash backtrace.
Stack dump:
0.      Program arguments: /opt/rocm/llvm/bin/llc /tmp/all_reduce-48c76c-gfx906-optimized-119b3c.bc -O2 -mtriple=amdgcn-amd-amdhsa -mcpu=gfx906 -filetype=obj -o /tmp/all_reduce-48c76c-gfx906-ad074c.o 
1.      Running pass 'CallGraph Pass Manager' on module '/tmp/all_reduce-48c76c-gfx906-optimized-119b3c.bc'.
2.      Running pass 'Module Verifier' on function '@_Z25ncclAllReduceTreeLLKernelILi2E7FuncSumIaEaEvP14CollectiveArgs'
 #0 0x000055e2f2d9022b llvm::sys::PrintStackTrace(llvm::raw_ostream&) (/opt/rocm/llvm/bin/llc+0x191b22b)
 #1 0x000055e2f2d8dde4 llvm::sys::RunSignalHandlers() (/opt/rocm/llvm/bin/llc+0x1918de4)
 #2 0x000055e2f2d8df39 (/opt/rocm/llvm/bin/llc+0x1918f39)
 #3 0x00007fad3ab09960 __restore_rt (/usr/lib/libpthread.so.0+0x14960)
 #4 0x00007fad3a59b355 raise (/usr/lib/libc.so.6+0x3c355)
 #5 0x00007fad3a584853 abort (/usr/lib/libc.so.6+0x25853)
 #6 0x000055e2f2d0d7e4 llvm::report_fatal_error(llvm::Twine const&, bool) (/opt/rocm/llvm/bin/llc+0x18987e4)
 #7 0x000055e2f2d0d909 (/opt/rocm/llvm/bin/llc+0x1898909)
 #8 0x000055e2f26e08a8 (/opt/rocm/llvm/bin/llc+0x126b8a8)
 #9 0x000055e2f2658ad0 llvm::FPPassManager::runOnFunction(llvm::Function&) (/opt/rocm/llvm/bin/llc+0x11e3ad0)
#10 0x000055e2f1eed3a1 (/opt/rocm/llvm/bin/llc+0xa783a1)
#11 0x000055e2f265a480 llvm::legacy::PassManagerImpl::run(llvm::Module&) (/opt/rocm/llvm/bin/llc+0x11e5480)
#12 0x000055e2f18afead main (/opt/rocm/llvm/bin/llc+0x43aead)
#13 0x00007fad3a586002 __libc_start_main (/usr/lib/libc.so.6+0x27002)
#14 0x000055e2f190d89e _start (/opt/rocm/llvm/bin/llc+0x49889e)
clang-11: error: unable to execute command: Aborted (core dumped)
clang-11: error: amdgcn-link command failed due to signal (use -v to see invocation)
clang-11: error: amdgcn-link command failed due to signal (use -v to see invocation)
clang-11: error: amdgcn-link command failed due to signal (use -v to see invocation)
clang-11: error: amdgcn-link command failed due to signal (use -v to see invocation)
clang version 11.0.0 
Target: x86_64-pc-linux-gnu
Thread model: posix
InstalledDir: /opt/rocm/llvm/bin
clang-11: note: diagnostic msg: Error generating preprocessed source(s).
make[2]: *** [CMakeFiles/rccl.dir/build.make:629: librccl.so.1.0] Error 254
make[2]: Leaving directory '/home/maverick/.cache/yay/rccl/src/build'
make[1]: *** [CMakeFiles/Makefile2:95: CMakeFiles/rccl.dir/all] Error 2
make[1]: Leaving directory '/home/maverick/.cache/yay/rccl/src/build'
make: *** [Makefile:171: all] Error 2
make: Leaving directory '/home/maverick/.cache/yay/rccl/src/build'
==> ERROR: A failure occurred in build().
    Aborting...
error making: %!s(func() string=0x55f1943e9bb0)
yay -S rccl  1828,47s user 246,03s system 332% cpu 10:23,94 total
```

**Comment by acxz** [comment]
For now if you aren't developing on the PKGBUILDs I would recommend to use the packages form `arch4edu`. See the [readme](https://github.com/rocm-arch/rocm-arch/blob/master/README.md) for instructions on how to use the binary packages arch4edu provides (be sure to give them a star!) if you just want to just install `rccl`. We only recommend compiling `rccl` if you plan on developing/contributing to the repository.

Although it is very weird that 3 people have been able to reproduce this issue.

**Comment by tpkessler** [comment]
It seems to be a compiler bug. I don't see how `nccl` may help here. Could you guys please provide more information on your system? Please post the output of `rocminfo` and `lscpu`. Do you use overclocked RAM? This sometimes triggers segfaults inside the compiler. Another possibility would be to build the newest `llvm-amdgpu` via the `llvm-amdgpu-git` package available on the AUR and try to build `rccl` with this compiler.

**Comment by acxz** [comment]
It has been a week and there has been no response/update. Since neither me nor @tpkessler can reproduce and the fact `rccl` is available on arch4edu, I'll go ahead and close this issue. Please feel free to comment on this thread with updated logs if the issue is still persisting and we will reopen this thread.

**Comment by rmsc** [comment]
I'm experiencing this exact same problem, rebuilding llvm-amdgpu-git right now.

I'm getting an out of resources error in rocminfo:

```
$ rocminfo
ROCk module is loaded
Unable to open /dev/kfd read-write: Cannot allocate memory
renato is member of render group
hsa api call failure at: /home/renato/.cache/yay/rocminfo/src/rocminfo-rocm-3.7.0/rocminfo.cc:1142
Call returned HSA_STATUS_ERROR_OUT_OF_RESOURCES: The runtime failed to allocate the necessary resources. This error may also occur when the core runtime library needs to spawn threads or create internal OS-specific events.
```

<details><summary>$ lscpu</summary>
<p>

```
Architecture:                    x86_64
CPU op-mode(s):                  32-bit, 64-bit
Byte Order:                      Little Endian
Address sizes:                   39 bits physical, 48 bits virtual
CPU(s):                          8
On-line CPU(s) list:             0-7
Thread(s) per core:              2
Core(s) per socket:              4
Socket(s):                       1
NUMA node(s):                    1
Vendor ID:                       GenuineIntel
CPU family:                      6
Model:                           60
Model name:                      Intel(R) Core(TM) i7-4720HQ CPU @ 2.60GHz
Stepping:                        3
CPU MHz:                         2385.539
CPU max MHz:                     3600,0000
CPU min MHz:                     800,0000
BogoMIPS:                        5189.70
Virtualization:                  VT-x
L1d cache:                       128 KiB
L1i cache:                       128 KiB
L2 cache:                        1 MiB
L3 cache:                        6 MiB
NUMA node0 CPU(s):               0-7
Vulnerability Itlb multihit:     KVM: Mitigation: VMX disabled
Vulnerability L1tf:              Mitigation; PTE Inversion; VMX conditional cache flushes, SMT vulnerable
Vulnerability Mds:               Mitigation; Clear CPU buffers; SMT vulnerable
Vulnerability Meltdown:          Mitigation; PTI
Vulnerability Spec store bypass: Mitigation; Speculative Store Bypass disabled via prctl and seccomp
Vulnerability Spectre v1:        Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:        Mitigation; Full generic retpoline, IBPB conditional, IBRS_FW, STIBP conditional, RSB
                                 filling
Vulnerability Srbds:             Mitigation; Microcode
Vulnerability Tsx async abort:   Not affected
Flags:                           fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts ac
                                 pi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc arch_perfm
                                 on pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf pni pclmulqdq dtes64
                                 monitor ds_cpl vmx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid sse4_1 sse4_2 x2apic mov
                                 be popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm cpuid_fault epb inv
                                 pcid_single pti ssbd ibrs ibpb stibp tpr_shadow vnmi flexpriority ept vpid ept_ad fsgs
                                 base tsc_adjust bmi1 avx2 smep bmi2 erms invpcid xsaveopt dtherm ida arat pln pts md_c
                                 lear flush_l1d
```
</p>
</details>

**Comment by rmsc** [comment]
I've just realized why rocminfo complains, my "old" GPU is not supported by rocm...

I'll finish the compilation of llvm-amdgpu-git though, and report back if it fixes the llvm crash.

**Comment by rmsc** [comment]
Couldn't build llvm-amdgpu-git, and not sure how to fix it:

```
/home/renato/.cache/yay/llvm-amdgpu-git/src/llvm-amdgpu-git/llvm/lib/Analysis/TFUtils.cpp:23:10: fatal error: tensorflow/c/c_api.h: No such file or directory
   23 | #include "tensorflow/c/c_api.h"
      |          ^~~~~~~~~~~~~~~~~~~~~~
compilation terminated.
```

**Comment by chengyuhui** [comment]
The `rccl` in `arch4edu` seems to be working (at least the files are there), but I'm being blocked by #413, which doesn't seem to be live on `arch`4edu`.

Building `llvm-amdgpu` is impossible for me, I don't have that amount of RAM on my system.

**Comment by acxz** [comment]
@rmsc the package has been updated to 3.8.0. Please let us know if you are still experiencing the issue.
As for llvm-amdgpu build errors please post them in a separate issue.

@chengyuhui rocblas has been updated to 3.8 on arch4edu.

I'll close this issue for now. Feel free to comment if the issue still persists.

**Comment by Zulleyy3** [comment]
Can reproduce this for 3.10.
[log.txt](https://github.com/rocm-arch/rocm-arch/files/5686571/log.txt)

**Comment by acxz** [comment]
@Zulleyy3 are you building in a clean chroot?

**Comment by tpkessler** [comment]
@Zulleyy3: Please provide more information on the packages you are using, in particular `llvm-amdgpu`. To me it seems like you're using its `git` version. Could you please either compile `llvm-amdgpu` (without `git`) or install its binary version from `arch4edu`?

Again, this is not a packaging error, but a compiler bug. Thus, it should be better reported upstream.

**Comment by Zulleyy3** [comment]
> @Zulleyy3 are you building in a clean chroot?
I am doing a clean build with yay

> Please provide more information on the packages you are using, in particular llvm-amdgpu.
I am using the `llvm-amdgpu` without git (Version 3.10.0-1) . I will try `arch4edu` in a bit.
I am also have `hip-rocclr` with version 3.10.0-1. `rocm-cmake` installed in the same version. 

PS: Is there a good way to get this info without specificially querying every package 1 by 1?

**Comment by tpkessler** [comment]
> I am using the llvm-amdgpu without git (Version 3.10.0-1)

Weird. I cannot reproduce this build error on two machines, both with a self-compiled `llvm-amdgpu` as well as with the binary version from `arch4edu`.

> Is there a good way to get this info without specificially querying every package 1 by 1?

You can run
```
pacman -Qo /opt/rocm
```

**Comment by Zulleyy3** [comment]
Ok, these are installed. The GPU in my System is an RX580. I will go try arch4edu now.

> /opt/rocm/ is owned by comgr 3.10.0-1
/opt/rocm/ is owned by hip-rocclr 3.10.0-1
/opt/rocm/ is owned by hsa-amd-aqlprofile-bin 3.10.0-1
/opt/rocm/ is owned by hsa-rocr 3.10.0-1
/opt/rocm/ is owned by hsakmt-roct 3.10.0-1
/opt/rocm/ is owned by llvm-amdgpu 3.10.0-1
/opt/rocm/ is owned by rocclr 3.10.0-1
/opt/rocm/ is owned by rocm-clang-ocl 3.10.0-1
/opt/rocm/ is owned by rocm-cmake 3.10.0-1
/opt/rocm/ is owned by rocm-dbgapi 3.10.0-1
/opt/rocm/ is owned by rocm-device-libs 3.10.0-1
/opt/rocm/ is owned by rocm-gdb 3.10.0-1
/opt/rocm/ is owned by rocm-opencl-runtime 3.10.0-1
/opt/rocm/ is owned by rocm-smi-lib64 3.10.0-1
/opt/rocm/ is owned by rocminfo 3.10.0-1
/opt/rocm/ is owned by rocprofiler 3.10.0-1
/opt/rocm/ is owned by roctracer 3.10.0-1

**Comment by acxz** [comment]
> I am doing a clean build with yay

I wonder if this is the reason why, maybe there is some package/environment flag out there that is interfering with the build.
If you can/have the time @Zulleyy3 I would love if you can build it in a clean chroot on your machine, since we can't really try to reproduce your entire build environment when building with yay. I guess if it doesn't build for you in a clean chroot I really don't know what the cause could be.

In the end though, I wouldn't worry much about it and just stick with the arch4edu, since they do the clean chroot building for you.

**Comment by Zulleyy3** [comment]
Using the llvm from arch4edu doesn't seem to help. If I have time, i will try building in a chroot.

**Comment by Zulleyy3** [comment]
I've been following this [](https://wiki.archlinux.org/index.php/DeveloperWiki:Building_in_a_Clean_Chroot) using pacman.conf that got it's build dependencies from the arch4edu repo.

It built. No issues whatsoever.

**Comment by acxz** [comment]
That's great to hear.

So there must be something in your (and in quite a few others) system that makes building rccl fail.

Considering that you should technically be building in a clean chroot or using arch4ed binaries anyway, I'll go ahead and close this. But if you do end up figuring out what the difference in your chroot and your actual system is which causes the build failure do feel free to comment on this thread.
